--- a/randr/rrscreen.c	2026-04-20 22:03:15.252079740 +0200
+++ b/randr/rrscreen.c	2026-04-20 22:37:34.277780016 +0200
@@ -278,7 +278,9 @@
             };
             pixman_f_transform_bounds(&crtc->f_transform, &display_box);
 
-            if (display_box.x2 > stuff->width || display_box.y2 > stuff->height)
+            /* if (display_box.x2 > stuff->width || display_box.y2 > stuff->height) */
+            if (crtc->x + mode->mode.width > stuff->width ||
+                crtc->y + mode->mode.height > stuff->height)
                 return BadMatch;
         }
     }
